File: java-spring/backend-integration-tests/src/test/java/net/chrisrichardson/eventstore/javaexamples/banking/backend/MoneyTransferIntegrationTest.java
package net.chrisrichardson.eventstore.javaexamples.banking.backend;

import io.eventuate.EntityWithIdAndVersion;
import io.eventuate.EventuateAggregateStore;
import net.chrisrichardson.eventstore.javaexamples.banking.accountsservice.backend.Account;
import net.chrisrichardson.eventstore.javaexamples.banking.accountsservice.backend.AccountService;
import net.chrisrichardson.eventstore.javaexamples.banking.common.transactions.TransferState;
import net.chrisrichardson.eventstore.javaexamples.banking.backend.common.transactions.TransferDetails;
import net.chrisrichardson.eventstore.javaexamples.banking.transactionsservice.backend.MoneyTransfer;
import net.chrisrichardson.eventstore.javaexamples.banking.transactionsservice.backend.MoneyTransferService;
import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.IntegrationTest;
import org.springframework.boot.test.SpringApplicationConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import java.math.BigDecimal;

import static net.chrisrichardson.eventstorestore.javaexamples.testutil.TestUtil.await;
import static net.chrisrichardson.eventstorestore.javaexamples.testutil.TestUtil.eventually;


@RunWith(SpringJUnit4ClassRunner.class)
@SpringApplicationConfiguration(classes = BankingTestConfiguration.class)
@IntegrationTest
public class MoneyTransferIntegrationTest {

  @Autowired
  private AccountService accountService;

  @Autowired
  private MoneyTransferService moneyTransferService;

  @Autowired
  private EventuateAggregateStore eventStore;


  @Test
  public void shouldTransferMoney() {
    final EntityWithIdAndVersion<Account> fromAccount = await(accountService.openAccount("00000000-00000000", "My Account", new BigDecimal(150), ""));

    final EntityWithIdAndVersion<Account> toAccount = await(accountService.openAccount("00000000-00000000", "My Account", new BigDecimal(300), ""));

    final EntityWithIdAndVersion<MoneyTransfer> transaction = await(
            moneyTransferService.transferMoney(new TransferDetails(fromAccount.getEntityId(),
                    toAccount.getEntityId(),
                    new BigDecimal(80))));


    eventually(
            () -> eventStore.find(Account.class, fromAccount.getEntityId()),
            account -> Assert.assertEquals(new BigDecimal(70), account.getEntity().getBalance()));

    eventually(
            () -> eventStore.find(Account.class, toAccount.getEntityId()),
            account -> Assert.assertEquals(new BigDecimal(380), account.getEntity().getBalance()));

    eventually(
            () -> eventStore.find(MoneyTransfer.class, transaction.getEntityId()),
            updatedTransaction -> Assert.assertEquals(TransferState.COMPLETED, updatedTransaction.getEntity().getState()));
  }

  @Test
  public void shouldFailDueToInsufficientFunds() {
    final EntityWithIdAndVersion<Account> fromAccount = await(accountService.openAccount("00000000-00000000", "My Account", new BigDecimal(150), ""));

    final EntityWithIdAndVersion<Account> toAccount = await(accountService.openAccount("00000000-00000000", "My Account", new BigDecimal(300), ""));

    final EntityWithIdAndVersion<MoneyTransfer> transaction = await(
            moneyTransferService.transferMoney(new TransferDetails(fromAccount.getEntityId(),
                    toAccount.getEntityId(),
                    new BigDecimal(200))));


    eventually(
            () -> eventStore.find(MoneyTransfer.class, transaction.getEntityId()),
            updatedTransaction -> Assert.assertEquals(TransferState.FAILED_DUE_TO_INSUFFICIENT_FUNDS, updatedTransaction.getEntity().getState()));

    eventually(
            () -> eventStore.find(Account.class, fromAccount.getEntityId()),
            account -> Assert.assertEquals(new BigDecimal(150), account.getEntity().getBalance()));

    eventually(
            () -> eventStore.find(Account.class, toAccount.getEntityId()),
            account -> Assert.assertEquals(new BigDecimal(300), account.getEntity().getBalance()));

  }


}

File: scala-spring/backend-integration-tests/src/test/scala/net/chrisrichardson/eventstore/examples/bank/MoneyTransferIntegrationTest.scala
package net.chrisrichardson.eventstore.examples.bank

import net.chrisrichardson.eventstore.EventStore
import net.chrisrichardson.eventstore.examples.bank.accounts.{AccountService, Account}
import net.chrisrichardson.eventstore.examples.bank.backend.common.accounts.AccountDebitedEvent
import net.chrisrichardson.eventstore.examples.bank.config.BankingTestConfiguration
import net.chrisrichardson.eventstore.examples.bank.queryside.AccountInfoQueryService
import net.chrisrichardson.eventstore.examples.bank.backend.common.transactions.TransferDetails
import net.chrisrichardson.eventstore.examples.bank.transactions.{MoneyTransferService, TransferStates, MoneyTransfer}
import net.chrisrichardson.eventstore.jdbc.config.JdbcEventStoreConfiguration
import org.junit.runner.RunWith
import org.scalatest.FlatSpec
import org.scalatest.junit.JUnitRunner
import org.scalatest.time.{Millis, Span}
import org.springframework.context.annotation.Import

import scala.concurrent.{Await, Future}
import scala.concurrent.duration._
import org.scalatest.concurrent.Eventually._
import org.scalatest.Matchers._
import scala.collection.JavaConversions._
import org.springframework.context.annotation.AnnotationConfigApplicationContext
import scala.reflect.ClassTag

trait SpringTest {
  def configClasses : Seq[Class[_]]
  val ctx = new AnnotationConfigApplicationContext(configClasses :_ *)
  def bean[T](implicit m : ClassTag[T]) : T = ctx.getBean(m.runtimeClass).asInstanceOf[T]
  def bean[T](name : String)(implicit m : ClassTag[T]) : T = ctx.getBean(name, m.runtimeClass).asInstanceOf[T]

}


@RunWith(classOf[JUnitRunner])
class MoneyTransferIntegrationTest extends FlatSpec with SpringTest {

  def configClasses = Seq(classOf[BankingTestConfiguration])

  val accountService = bean[AccountService]
  val transactionService = bean[MoneyTransferService]
  val eventStore = bean[EventStore]
  val accountInfoQueryService = bean[AccountInfoQueryService]

  def await[T](body : => Future[T]) = Await.result(body, 500 milliseconds)

  val longWait = PatienceConfig(Span(5000, Millis), Span(50, Millis))


  it should "transfer money" in {
    val account1 = await { accountService.openAccount(BigDecimal(150)) }
    val account2 = await { accountService.openAccount(BigDecimal(300)) }

    val transaction = await { transactionService.transferMoney(TransferDetails(account1.entityId, account2.entityId, BigDecimal(80))) }
    
    eventually {
      val updatedTransaction = await { eventStore.find[MoneyTransfer](transaction.entityId)}
      updatedTransaction.entity.state should be(TransferStates.COMPLETED)
    }(longWait)

    eventually {
        val updatedAccount1 = await { eventStore.find[Account](account1.entityId) }
        updatedAccount1.entity.balance should be(70)
    }(longWait)

    eventually {
        val updatedAccount2 = await { eventStore.find[Account](account2.entityId) }
        updatedAccount2.entity.balance should be(380)
    }(longWait)

    eventually {
      val accountInfo = accountInfoQueryService.findByAccountId(account1.entityId)
      accountInfo.balance should be(70*100)
      accountInfo.changes.map(_.transactionType) should be(Seq(classOf[AccountDebitedEvent].getSimpleName))
    }(longWait)

  }

  it should "fail to transfer money due to insufficient funds" in {
    val account1 = await { accountService.openAccount(BigDecimal(150)) }
    val account2 = await { accountService.openAccount(BigDecimal(300)) }

    val transaction = await { transactionService.transferMoney(TransferDetails(account1.entityId, account2.entityId, BigDecimal(200))) }

    eventually {
      val updatedTransaction = await { eventStore.find[MoneyTransfer](transaction.entityId)}
      updatedTransaction.entity.state should be(TransferStates.FAILED_DUE_TO_INSUFFICIENT_FUNDS)
    }(longWait)

    eventually {
      val updatedAccount1 = await { eventStore.find[Account](account1.entityId) }
      updatedAccount1.entity.balance should be(150)
    }(longWait)

    eventually {
      val updatedAccount2 = await { eventStore.find[Account](account2.entityId) }
      updatedAccount2.entity.balance should be(300)
    }(longWait)

  }

}
